Add High DPI support

AustinWise opened the pull request:
# Background

Computers with higher pixel density are becoming more common. These screens have more dots-per-inch (DPI) than display in that past typically had. For applications built on top of the traditional Win32 API (including this Windows Forms base application), the application has to explicitly support higher DPI. If an application does not support high DPI and is run on a high DPI system, Windows will run the application at the standard DPI and upscale the resulting pixels. The result is an application which looks blurry.

# This change

This changes does some basic changes to support high DPI, including changing the manifest and changing the scale modes of forms and controls.

# Outstanding items that prevent being considered for merge

* [x] Run unit tests at a consistent DPI
* [x] Make unit tests pass
  * The `TestDragDrop` test does not work on my PC on the master branch and this PR.
* [x] Ensure app works when DPI is changed while it is running
* [x] Ensure app works correctly when DPI is changed while app is *not* running
  * Specifically make sure the font size setting is persisted and read correctly
* [x] Test with multiple scaling factors besides 150%
* [x] Ensure all forms and dialogs are using `AutoScaleMode = System.Windows.Forms.AutoScaleMode.Dpi`
* [ ] Scale scroll bars width automatically
* [ ] Scale size of icons in the tree view

AustinWise commented:
Currently this change does not pass unit tests. It pasts most of them when run with low DPI (commenting out the `System.Windows.Forms.ApplicationConfigurationSection` section in app.config). It fails more when run on high DPI.

I think the remaining work that needs to be done is:

1. Figure out how to run the application in a consistent DPI mode during unit tests, so unit tests won't have to add high-DPI support.
2. Fix the unit tests.

If there is interest in this sort of change, I can work on resolving these issues.

AustinWise commented:
Here is how it looks at 150% scaling:

![image](https://user-images.githubusercontent.com/7751/111894779-dda66200-89ca-11eb-99c0-a0a6635e0298.png)

lovettchris commented:
This is great, thanks so much for doing this, and the winforms generation compile error fix is also super awesome - that's been bugging me for ages.

Hey, I've tried your branch on my machine with my system set to 150% and I see now that XML notepad shrinks back to normal size when screen is in high DPI so old is big (front) and new is normal size (back):

![image](https://user-images.githubusercontent.com/18707114/112043178-9ffa2400-8b05-11eb-9614-b65a4830af24.png)

My question then is if someone is setting high DPI for accessibility reasons (they need to zoom the size of everything in order to be able to read it) then the new behavior may not be what they want.  Perhaps then this needs to be a Setting in the Options dialog where a user can choose to scale or not...   I notice most other apps, including the web browser do not do what you are proposing for XML notepad...

I also see some layout bugs in XML notepad when I dynamically change the DPI from 100 to 150 the address bar doesn't resize properly.

Also should all XML notepad dialogs be setting "this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Dpi" in order to be consistent?

AustinWise commented:
I've updated the description of the PR with the issues you have brought up. There is still work to be done to make things work correctly when the DPI changes while the app is running. Besides the problem with the address bar position, I've noticed that sometimes the font size get changed when changing DPIs, which results in tiny text.

I *think* the problems you are showing in the screen shot are related to changing the scaling setting. When I log in with 150% scaling and run both the release version of the application and my patched version, with the standard 8.25pt font setting, they look pretty similar:

![image](https://user-images.githubusercontent.com/7751/112095654-1c1e5700-8b5a-11eb-9ec5-a71f402ae655.png)

I will take a look at resolving these issues this weekend. At that time I will also break out the designer fix into a separate PR, since it is a nice quality-of-life fix that everyone can benefit from, regardless of the fate of this PR.

lovettchris commented:
Awesome thanks, yeah the separate fix PR would be great, I'll merge that one as soon as it is ready.

AustinWise commented:
@lovettchris I made a separate PR just for the designer-generated code issue: #74

lovettchris commented:
Thanks, and I see how that autoscaling is making it look the same as the old version of xmlnotepad only the fonts are much more clear and less blurry.  Cool.

![image](https://user-images.githubusercontent.com/18707114/112404787-3da04f80-8cce-11eb-8dd2-f3045c62bd8c.png)

lovettchris commented:
I had a closer look at failing unit tests under DPI scaling, and fixed some of the more obvious bugs.  See the changes in this PR: https://github.com/microsoft/XmlNotepad/pull/79.  But there's still some mysterious behavior with mouse clicks that go off to unexplainable locations on the screen and one some times, not all the time, very strange...

AustinWise commented:
I fixed the layout and font size persist when the DPI is changed while the app is running. Windows Forms keeps the app the same size after changing the DPI, so you have to restart the app for the new DPI to truly take effect.

Not all the unit tests passed with the Windows Forms DPI scaling, so I added support for an environmental variable to disable Windows Forms DPI scaling and use the system scaling instead. This fixed the unit tests. I have not looked too deeply into why the unit tests don't work with the DPI scaling enabled. Obviously it would be nice for the unit tests to work at any DPI, but that might be a lot of work.

The only things not scaling are the icons in the tree view and the scroll bar widths.

lovettchris commented:
Hey man you've done great, I can take care of the icons.  let me know if you are done and I'll merge this PR.

AustinWise commented:
I think I might not get a chance to do the icon rendering soon. For my purposes the application is just as usable as before with clearer text, so feel free to merge if you would like.

lovettchris commented:
@AustinWise Your awesome work is published in the new version 2.8.0.28.  I also adding some layout fixes so the window now updates properly when you move it from a monitor with DPI 100% to a monitor with DPI 150% and back.  Thanks again!
